discuss tradeoffs of mocking and faking

maxheld83 opened the pull request:
closes #90

maxheld83 commented:
btw, I love that you guys are hard-wrapping by sentence in the rmd ☺️. 
Been trying to lobby the tidyverse to adopt that elsewhere: https://github.com/tidyverse/style/issues/162

maelle reviewed (commented):
Thank you!! Two remarks

* I like the content but feel it should come later in the book are things are complicated enough at this stage. Maybe a compromise would be to have a sentence in the intro (or conclusion of the Whole Games section) in a very small section "Can one use packages for HTTP testing as little as possible?" or so and have a link to a chapter later in the book, that'd feature this current content (in advanced topics). 

* In your own packages, are there examples of code or tests that you could permalink and comment to help make things less theoretical? Or even false code? E.g. when you write "other functions", you could explain what those do e.g. data munging, what the functions interacting with the API return, to help readers visualize things a bit more?

> btw, I love that you guys are hard-wrapping by sentence in the rmd relaxed.

Hehe that was my good resolution one year, thanks to Gábor Csárdi's commenting on this once in a R-hub post review. It's the last item in rOpenSci blog guidelines: https://blogguide.ropensci.org/technical.html#styleguide

maxheld83 commented:
Great + agreed! 
I'll work in those suggestions @maelle -- could be a day or two.

maelle commented:
thanks, no hurry!

sckott commented on intro-pkgs.Rmd:
```diff
@@ -21,6 +21,31 @@ What are the challenges of HTTP testing?
+Both approaches let you run your tests against a state of an external HTTP API which you fully control.
+This can be a good thing, because it allows you to test against all theoretically possible behavior of the API, as opposed to just the current behavior of the live API.
+
+More fundamentally, by taking control of the API behavior, your can test your code, and only your code.
```
> taking control of the API behavior

can you clarify this?

maxheld83 replied:
I just got rid of this convoluted language.

sckott commented on intro-pkgs.Rmd:
```diff
@@ -21,6 +21,31 @@ What are the challenges of HTTP testing?
+This tradeoff cannot be resolved, but it is eased considerably by decoupled (or modular) code.
+Much as with other side effects, it can help to concentrate your API calls in as few functions as possible and test these with mocks or fakes, where necessary.
+All other functions can then be tested without dependence on the API.
+Conversely, if you find yourself requiring these test doubles in a lot of places, you may want to [consider a refactoring](https://medium.com/javascript-scene/mocking-is-a-code-smell-944a70c90a6a).
```
"test doubles" should be defined i think

maxheld83 replied:
done

sckott commented on intro-pkgs.Rmd:
```diff
@@ -21,6 +21,31 @@ What are the challenges of HTTP testing?
+
+This tradeoff cannot be resolved, but it is eased considerably by decoupled (or modular) code.
+Much as with other side effects, it can help to concentrate your API calls in as few functions as possible and test these with mocks or fakes, where necessary.
+All other functions can then be tested without dependence on the API.
```
Perhaps a tradeoff should be discussed for this point: I assume not testing a function against an API means testing it against some kind of test double or do you mean something else? And test doubles can at some point in the future not reflect what the API currently returns. In `vcr` this is what `re_record_interval`   https://docs.ropensci.org/vcr/reference/vcr_configure.html?q=interval#arguments attempts to achieve (TBH though i've not used this very much, mostly b/c I've not got around to setting it up)

maxheld83 replied:
I've clarified the language and example (hopefully)

maelle commented:
I was just thinking about proper attribution of work @maxheld83

* Please add a sentence to the beginning of the chapter you'll add, mentioning it was contributed by you.
* Add a sentence just before https://github.com/ropensci-books/http-testing/blob/13996fbfeabde2975acb054249106608931bf898/index.Rmd#L31 (we'll now start listing contributor as in https://devguide.ropensci.org/preface.html)

maxheld83 commented:
yikes, sorry I took so long to follow up:

> I like the content but feel it should come later in the book are things are complicated enough at this stage. Maybe a compromise would be to have a sentence in the intro (or conclusion of the Whole Games section) in a very small section "Can one use packages for HTTP testing as little as possible?" or so and have a link to a chapter later in the book, that'd feature this current content (in advanced topics).

- [x] The discussion now lives in its own chapter, with a short pointer in the beginning.
 
> In your own packages, are there examples of code or tests that you could permalink and comment to help make things less theoretical? Or even false code? E.g. when you write "other functions", you could explain what those do e.g. data munging, what the functions interacting with the API return, to help readers visualize things a bit more?

- [x] I've added an example of how to minimise / tradeoff mocking/faking from a real-life package (http://subugoe.github.io/biblids/).
    I've simplified the code a bit, but it works.

- [x] added attribution as requested.

maelle commented on conclusion.Rmd:
```diff
@@ -1,5 +1,3 @@
-#  (PART) Conclusion {-}
```
your chapter should live in "advanced topics" I think. It'd make the TOC clearer.

maxheld83 replied:
done

maelle commented on intro-pkgs.Rmd:
```diff
@@ -113,5 +113,8 @@ Now that you have an idea of the tools we can use for HTTP testing, we'll now cr
 
 Our minimal package will use httr. However, it will help you understand concepts even if you end up using crul or curl.[^limits]
 
+Test doubles such as mocks and fakes are helpful tools, but they also introduce their own complexity.
```
"programs design" is a bit unclear.

maelle replied:
I think the mention of the chapter should best happen in the conclusion as it might be confusing and overwhelming in the intro.

maelle replied:
I know I said otherwise earlier. :thinking: Maybe the paragraph should be much shorter and not mention code design / code smell as it makes it look potentially scary right from the intro.

maxheld83 replied:
I think there's a very real risk here to overbuild and overcomplicate the testing, while missing deeper flaws.
So I think a warning is warranted.

maelle replied:
fair enough!

maelle commented on tradeoffs.Rmd:
```diff
@@ -0,0 +1,172 @@
+:::
+
+Test doubles such as mocks or fakes can help you isolate your tests from the state of external HTTP API.
+This can be a good thing, because it allows you to test your code against a rare responses, or when the API is not available.
```
you can link to https://books.ropensci.org/http-testing/pkgs-testing-chapter.html#why-do-we-need-special-packages-for-http-testing `#pkgs-testing-chapter`

maxheld83 replied:
done!

maelle commented on tradeoffs.Rmd:
```diff
@@ -0,0 +1,172 @@
+Test doubles such as mocks or fakes can help you isolate your tests from the state of external HTTP API.
+This can be a good thing, because it allows you to test your code against a rare responses, or when the API is not available.
+
+Unfortunately, mocking and faking add their own complexity to a project.
```
But fortunately this book makes it very clear how to :stuck_out_tongue_winking_eye:

maxheld83 replied:
but complexity still remains ... complex 🙃

maelle commented on tradeoffs.Rmd:
```diff
@@ -0,0 +1,172 @@
+You now have two relationships to keep in mind: your code must pass your mocked/faked tests, and your mocks/fakes must correspond to the live API.
+
+This tradeoff cannot be resolved, but it is eased considerably by decoupled (or modular) code.
+Much as with other side effects, it can help to concentrate your API calls in as few functions as possible and test these with mocks or fakes, where necessary.
```
Isn't this a repetition of https://books.ropensci.org/http-testing/graceful.html#graceful-code?

Could you add some mock tests (ahah)? I.e. two functions calling an API, a version without and a version with utility functions?

maxheld83 replied:
> Isn't this a repetition of https://books.ropensci.org/http-testing/graceful.html#graceful-code?

Yes, sorry -- added a reference.
I think the angle here is somewhat different, drawing the parallel to side effects, which we generally want to avoid and minimise/concentrate in as few functions as possible.

maxheld83 replied:
> Could you add some mock tests (ahah)? I.e. two functions calling an API, a version without and a version with utility functions?

I'm not sure I fully understand.
I think the hypothetical below `get_doi_url()` might be this example, where I mention that it (and other upstream functions) won't need to query the API themselves, but can rely on `get_doi_handle()` instead.
I guess this is the kind of thing that would benefit from integrating the examples with all of the text as you mentioned.

maelle replied:
oops I meant mock code not mock tests!!

maelle commented on tradeoffs.Rmd:
```diff
@@ -0,0 +1,172 @@
+
+This tradeoff cannot be resolved, but it is eased considerably by decoupled (or modular) code.
+Much as with other side effects, it can help to concentrate your API calls in as few functions as possible and test these with mocks or fakes, where necessary.
+All other functions can then be tested without dependence on the API.
```
It might make sense to be less sure e.g. "it might help with not having to test all functions with the API" or so.

maxheld83 replied:
Isn't "can help" already careful?
I like simple, opinionated heuristics.

maelle commented on tradeoffs.Rmd:
```diff
@@ -0,0 +1,172 @@
+This tradeoff cannot be resolved, but it is eased considerably by decoupled (or modular) code.
+Much as with other side effects, it can help to concentrate your API calls in as few functions as possible and test these with mocks or fakes, where necessary.
+All other functions can then be tested without dependence on the API.
+Conversely, if you find yourself requiring these test doubles in a lot of places, you may want to [consider a refactoring](https://medium.com/javascript-scene/mocking-is-a-code-smell-944a70c90a6a).
```
Instead of an external link or as a complement to it I'd add an example or maybe if you add the mock code earlier it is enough.

maxheld83 replied:
I pointed to the below example here now, but again, would benefit from the full integration with the examples you suggested.

maelle commented on tradeoffs.Rmd:
```diff
@@ -0,0 +1,172 @@
+On the other hand, [*integration tests*](https://en.wikipedia.org/wiki/Integration_testing) should test how your code interacts with other software or services.
+["Narrow" integration tests](https://martinfowler.com/bliki/IntegrationTest.html) are run against test doubles, typically fakes.
+This may be most helpful to test more complex, perhaps stateful, interactions between your code and the API, which relies on internal logic of the API rather than a one-off response.
+For example, you may want to test whether if you submit a string to a translation service, you get back a result in the appropriate language.
```
Citing these notions is interesting for sure, but it would help to link to real R packages using them otherwise the chapter might be too dry, what do you think?

maxheld83 replied:
I take up these notions in the below example, and I think it's just helpful vocabulary.
I'd hope that you can find these distinctions in some R packages, though I wasn't able to find a clean-cut example. 
Also seems to me that explaining the context of another API wrapper might bloat the chapter more.

maelle replied:
But even in a few words, what would be a narrow integration test for an R package?

maxheld83 replied:
ah, I see, yes, that sort of illustration really is needed.

maelle commented on tradeoffs.Rmd:
```diff
@@ -0,0 +1,172 @@
+For example, you may want to test whether if you submit a string to a translation service, you get back a result in the appropriate language.
+Such test duplication of external business logic may be hard to get right and maintain, and should rarely be necessary.
+
+["Broad" integration tests](https://martinfowler.com/bliki/IntegrationTest.html) are run against live services, and thus require no test doubles.
```
You could link to https://books.ropensci.org/http-testing/real-requests-chapter.html :slightly_smiling_face:

maxheld83 replied:
done!

maelle commented on tradeoffs.Rmd:
```diff
@@ -0,0 +1,172 @@
+As a rule of thumb, your code should handle, and your tests should cover the failure *modes* of the API, including HTTP response codes, API response codes or incorrect response formats.
+Perhaps you can additionally test some API requests against some known outputs, but this will always remain a sample.
+There's always a chance that some deeply nested JSON response will break your code.
+You may not be able to cover all such errors from an errand or changed API, and that's okay.
```
add some sort of hopeful sentence i.e. you'd fix the code and add a test for that when you see such an error.

maelle commented on tradeoffs.Rmd:
```diff
@@ -0,0 +1,172 @@
+There's always a chance that some deeply nested JSON response will break your code.
+You may not be able to cover all such errors from an errand or changed API, and that's okay.
+
+## Example
```
I like seeing an example and wonder how it could be more coupled with the theoretical concepts from earlier in the chapter. I.e. having the example first, with notions explained with it.

maxheld83 replied:
My idea was to focus:

1) non-example part on *why* you'd want to minimise your API exposure
2) how to minimise that in a toy example

Might also work well if it is restructured as you suggest, but I don't think I can do that now -- maybe first resolve overall question of tone and direction.

maelle commented on tradeoffs.Rmd:
````diff
@@ -0,0 +1,172 @@
+
+```{r minimal-wrapper}
+# unexported helper
+# this is not tested, b/c the implementation could change
````
why is it not tested?

maelle replied:
I mean it's tested _indirectly_?

maxheld83 replied:
Hopefully explained better what I meant: internal functions aren't supposed to have tests.

maxheld83 replied:
I debated adding roxygen tags to the example, but that just bloated it too much and is a separate concern anyway.

maelle commented on tradeoffs.Rmd:
```diff
@@ -0,0 +1,172 @@
+verb_doi_handle <- function(doi, ...) {
+  httr::VERB(
+    url = "https://doi.org/",
+    # better to use curl::curl_escape() here
```
maybe this is not needed here?

maxheld83 replied:
done

maelle commented on tradeoffs.Rmd:
```diff
@@ -0,0 +1,172 @@
+  httr::stop_for_status(resp) # covers outcomes 1 *and* 3 from above list
+  if (httr::content(resp)$responseCode == 200) {
+    # covers outcome 2b
+    stop("The handle exists, but the queried value(s) could not be found.")
```
By the way one could imagine your function would not fail but return a NULL instead or some data.frame with some sort of explanatory column?

I can see the next function does something linked to that but it'd mean you call the API twice? Am I missing something?

maxheld83 replied:
Not sure I fully understand.

Do you mean the function *should* return a `NULL` or `data.frame()` with some output if the queried value does not exist?

I haven't fully thought this through, but I like functions to fail early and to maintain type/length-stability at almost any cost, which is hard to do if you build these kind of special cases ...

But maybe I misunderstood your comment ...

maelle commented on tradeoffs.Rmd:
```diff
@@ -0,0 +1,172 @@
+})
+# outcome 2a
+test_that("Existing DOIs are found", {
+  testthat::local_edition(3)
```
I'd argue this adds to much complexity that's unrelated to the matter of the chapter?

maxheld83 replied:
yes indeed!

maelle commented on tradeoffs.Rmd:
````diff
@@ -0,0 +1,172 @@
+```
+
+One might think that these tests should be mocked, rather than running against the live API.
+However, because DOI.org is stable, performant and requires no authentication, the added complexity is not worth it in this case.
````
What do the API maintainers say about this btw? 
```suggestion
However, because DOI.org is stable, performant and requires no authentication, the added complexity is not worth in my opinion it in this case.
```

maxheld83 replied:
Their [documentation](https://www.doi.org/factsheets/DOIProxy.html#rest-api) is *very* sparse 🙂.
But I think it's backed by a proxy or CDN or some such thing, and it doesn't even *have* auth, so I'd think these relatively few requests (dozends) per test run are ok.

maxheld83 replied:
the "in my opinion" seems a bit superfluous, no?

maelle commented on tradeoffs.Rmd:
````diff
@@ -0,0 +1,172 @@
+})
+```
+
+Happily, we can cheat a little, and simply create *some* sure HTTP status error, without actually reproducing API behavior in any detail.
````
the whole book is about happily cheating :grin:

maxheld83 replied:
I think it's very rigorous 😊

maelle commented on tradeoffs.Rmd:
```diff
@@ -0,0 +1,172 @@
+Of course, if `get_doi_url()` did some very complicated parsing of the API response (it doesn't), we might want to test that behavior, too.
+But in that case, we should let it accept the API response as a parameter and test that as we would any other function free of side effects.
+
+## Use Your Judgement
```
I think it'd also be good to add something about a team of developers: it might help, to find the balance, to have others review your code, to exchange ideas etc. Even just to speak with others at conferences etc. (maybe I'm saying that because I watched your interesting useR! talk :slightly_smiling_face: ).
Adding such a sentence about community here would IMHO make it look less unattainable and scholarly?

maxheld83 replied:
I think that just bloats the chapter.
Asking for help and getting fresh eyes applies to pretty much everything, no?

maelle reviewed (changes requested):
Thanks a lot @maxheld83, that's a very interesting discussion to have!

My main comments are around
- moving this to Advanced Topics rather than the conclusion;
- make the reference to the chapter in the introduction less overwhelming;
- make the example more central so that the theory looks less dry;
- adding references to other chapters of the book.

Thanks a ton!

maxheld83 commented:
thanks so much for taking the time to so carefully review this again @maelle!

I've fixed some of the smaller things and changed the chapter location.

But I'm not sure about the overall tone and direction.

I appreciate that the book is trying to be very hands-on and welcoming, but I think my concern here is/was indeed that there's a very real danger to jump into the tools *before* carefully thinking through the concepts.
I prefer to first worry about the concepts (the purpose of TDD in this case?), and then jump into the tooling.
I don't have any formal training, and  I think I've in the past often excitedly jumped onto new tools and built myself into a crazy complicated corner.

I also personally find opinionated heuristics ("strong opinions weakly held") helpful, though that's a matter of style. 
(And I may just be *wrong*).

I wonder whether this chapter actually fits into the book, because of these (potential) differences.
This is your show after all, and a book should speak with one voice/tone, so I'd be happy to move this over to biblids or a blog, and then perhaps a short link to that can point people to it, if they're curious.

Let me know how you feel.

maelle commented:
Thank you! Maybe you could publish it as a blog post first, and then, once you've received a bit of feedback, come back to adding it as a chapter here if you so wish? :thinking:

maelle commented:
(I'll be happy to link to the blog post in any case, as again, this is important :-) )

maxheld83 commented:
> Thank you! Maybe you could publish it as a blog post first, and then, once you've received a bit of feedback, come back to adding it as a chapter here if you so wish

That's a great idea.
I'll let this linger as a "draft" then until there's some feedback on another venue.

maelle commented:
Closing for lack of recent activity (but please ping me if you wrote a blog post, and thanks a lot for contributing!)
